README.md
# Prosopopee

More or less a small clone of exposure.co in form of a static generator. For
those of you who don't know what exposure.co is, this more or less allows you
to organise you picture in a way that more or less tell a story.

You can find example usages here:

* http://media.faimaison.net/photos/galerie/
* https://www.thebrownianmovement.org/
* http://outside.browny.pink
* http://such.life

# Why?

I wanted to learn a bit of advanced css and I wanted to self-host my data.

## Requirements

Installation needs Python, pip and virtualenv

    apt-get install python-pip python-virtualenv

Gallery building needs graphicsmagick library

    apt-get install graphicsmagick

## Installation

1. Create a virtualenv, and activate it

```
virtualenv ve
source ve/bin/activate
```

2. Download and install Prosopopee

```
pip install prosopopee
```

## Files organisation

The files organisation is quite simple:

* in the root directory of your project you need a settings.yaml file that will contains the title and subtitle of your gallery
* for each gallery you'll need a folder that also contains a settings.yaml file that will describe how to display the content on your gallery
* and you put the pictures of the gallery inside the gallery folder

### Root settings.yaml

The root settings.yaml should contains 2 keys : one for the title of your website and one for the subtitle. It should looks like that:

```yaml
title: My exploration of the outside world
sub_title: it's a scary place, don't go there
```

It can also optionally contain a menu and global settings.

#### Menu

It is possible to add a menu to your homepage that links to static pages. To do so, add a `menu` key to your `settings.yaml`, for example:

```yaml
menu:
  - about: "About"
  - first_gallery: "My first gallery"
  - http://twitter.com: "Twitter"
```

For example, this could be the content of `settings.yaml` in `about` folder:

```yaml
title: "About"
static: true
public: false
sections:
  - type: text
    text: Some text, HTML <b>is allowed</b>.
```

You can use the `static` option to get a template closer to the one of the
homepage that is better suited for a static page. You'll need to specify
"public: false" if you don't want to list this page on the homepage. On you
case you didn't specified "public: false" you'll **need** to specify a "cover:"
entry like any other gallery.

**NOTE**: except the "static: " option to disepear quite soon for a more
generic approach to "choose your page style".

#### Global settings

Global settings can be set in your root `settings.yaml`, under the `settings` key.

##### GM

Currently a `gm` settings key allows to customize the default GraphicsMagick's behavior. It looks like :

```yaml
settings:
  gm:
    quality: 75
    auto-orient: True
    strip: True
    resize: 50%
```

The meaning of the currently supported GraphicsMagick's settings is as follows :
* `quality` allows to customize the compression level of thumbnails (between 0 and 100)
* `auto-orient` change the orientation of pictures so they are upright (based on corresponding EXIF tags if present)
* `strip` removes all profiles and text attributes from the image (good for privacy, slightly reduce file size)
* `resize` can be used to resize the fullsize version of pictures. by default, input image size is preserved

Any GraphicsMagick setting can be customized on a per-image basis (either `cover` or `image`, see below).

##### Themes

Prosopopée has a support for various themes. As for now, only 2 themes are available:

* the default one called "exposure"
* "material" based on materialcss

To specify the theme, add the "theme" key in your "settings" key or your
**root** settings.yaml. For example:

```yaml
title: My exploration of the outside world
sub_title: it's a scary place, don't go there
settings:
  theme: material
```

### Gallery settings.yaml

This settings.yaml will describe:

* the title, subtitle and cover picture of your gallery that will be used on the homepage
* the tags is optional
* if your gallery is public (if not, it will still be built but won't appear on the homepage)
* the date of your gallery: this will be used on the homepage since **galleries are sorted anti chronologically** on it
* the list of sections that will contains your gallery. A section will represent either one picture, a group of pictures or text. The different kind of sections will be explained in the next README section.

Here is an example:

```yaml
title: Gallery title
sub_title: Gallery sub-title
date: 2016-01-15
cover: my_cover_picture.jpg
tags:
  - #yolo
  - #travel
sections:
  - type: full-picture
    image: big_picture.jpg
    text:
      title: Big picture title
      sub_title: Some text
      date: 2016-01-15
  - type: pictures-group
    images:
      -
        - image1.jpg
        - image2.jpg
        - image3.jpg
      -
        - image4.jpg
        - image5.jpg
  - type: text
    text: Some text, HTML <b>is allowed</b>.
  - type: bordered-picture
    image: another_picture.jpg
```

And here is an example or a **private** gallery (notice the <code>public</code> keyword):

```yaml
title: Gallery title
sub_title: Gallery sub-title
date: 2016-01-15
cover: my_cover_picture.jpg
public: false
sections:
    - ...
```

#### Images handling

Images go into the `cover` or `image` keys.
Each image individual processing settings can be customized to override the default
GraphicsMagick settings defined (or not) in the root `settings.yaml`.

This is done by putting the image path into a `name` key,
and adding specific processing settings afterwards.

For example, you can replace :

```yaml
image: image1.jpg
```

by:

```yaml
image:
  name: image1.jpg
  quality: 90
  strip: False
  auto-orient: False
```


### Different kind of sections

A gallery is composed of a succession of sections as you can see on this [wonderfully
totally uninteresting example
gallery](http://psycojoker.github.io/prosopopee/first_gallery/) the gallery is
composed of 5 sections:

* a full screen picture with text written on it
* a picture with borders around it
* a group of 5 pictures
* and a fullscreen picture without text on it this time

In your settings.yaml, a section will **always** have a <code>type</code> key
that will describe its kind and additional data. Underneath, the
<code>type</code> key is actually the name of an HTML template and the other
data will be passed to this template.

You can find all the sections templates here: https://github.com/Psycojoker/prosopopee/tree/master/prosopopee/templates/sections

You often have an <code>image</code> key. You need to give it a path to the
actual file. By convention, those files are put inside your gallery folder but
this is not mandatory.

#### Full Screen picture with OR without text on it

This displays a full screen picture as shown in the [example
gallery](http://psycojoker.github.io/prosopopee/first_gallery/) in the first
and last sections. How you should use it:

With text:

```yaml
  - type: full-picture
    image: big_picture.jpg
    text:
      title: Big picture title
      sub_title: Some text
      date: 2016-01-15
```

Without text:

```yaml
  - type: full-picture
    image: big_picture.jpg
```

#### Bordered picture

This displays a centered picture that is surrounded by white (the background) as
shown in the second position of the [example
gallery](http://psycojoker.github.io/prosopopee/first_gallery/).

How to use it:

```yaml
  - type: bordered-picture
    image: another_picture.jpg
```

#### Group of pictures

This displays a group of zoomable pictures on one or multiple lines as shown on
the fourth position (after the text) of the [example
gallery](http://psycojoker.github.io/prosopopee/first_gallery/).

```yaml
  - type: pictures-group
    images:
      -
        - image1.jpg
        - image2.jpg
        - image3.jpg
      -
        - image4.jpg
        - image5.jpg
```

The first level `-` represent a line of pictures.
The second level `-` represent the list of images in this line.

**Know bug**: the images are left aligned, so if you don't put enough images on
a line, you'll have white space on the right.

#### Text

This displays some centered text as shown on the third position of the [example
gallery](http://psycojoker.github.io/prosopopee/first_gallery/). HTML is
allowed inside the text.

How to use it:

```yaml
  - type: text
    text: Some text, HTML <b>is allowed</b>.
```

#### Paragraph

This displays a h2 title followed by text. HTML is allowed inside of the text.
If no title is declared, a separator is added.

How to use it:

```yaml
  - type: paragraph
    title: the title
    text: Some text, HTML <b>is allowed</b>.
```

#### HTML

This section is for raw html that will be centered (for example: inlining an OSM iframe).

How to use it:

```yaml
  - type: html
    html: <tag>some html stuff</html>
```

#### Panorama

This displays a very large picture with a drag-and-drop possibility on it.

How to use it:

```yaml
  - type: panorama
    image: 7.jpg
```

### Example

As a recap, here is how the files of the example gallery are organised:

```
example
      ├── settings.yaml
      └── first_gallery
          ├── settings.yaml
          └── stuff.png
```

The content of <code>example/settings.yaml</code>:

```yaml
title: "Example gallery"
```

The content of <code>example/first_gallery/settings.yaml</code>:

```yaml
title: my first gallery
sub_title: some subtitle
date: 2015-12-08
cover: stuff.png
sections:
  - type: full-picture
    image: stuff.png
    text:
      title: Beautiful Title
      sub_title: pouet pouet
      date: 2015-12-08
  - type: bordered-picture
    image: stuff.png
  - type: text
    text: « voici plein de blabla à rajouter et <b>ceci est du gras</b> et encore plein plein plein plein de text car je veux voir comment ça va wrapper car c'est important et il faut pas que j'oublie de mettre des margins en % sinon ça va pas le faire alala là ça devrait aller »
  - type: pictures-group
    images:
      -
        - stuff.png
        - stuff.png
        - stuff.png
      -
        - stuff.png
        - stuff.png
  - type: full-picture
    image: stuff.png
```

## Build the website

**Note: You need to be in an activated virtualenv.**

In a folder containing the **root** settings.yaml file, simply do

    prosopopee

A `build` folder will be created in the current directory, containing an
index.html, static files (css & js) and pictures.

## TODO

* write documentation on how to overwrite the templates
* allow to overwrite static files (for that we need a merge mechanism)
* be mobile responsivepouetpouet (this mean css + resize pictures)
* probably moar sections + allows to add legends on pictures or something like that
* add `--source` & `--destination` command line options
* add rss/atom

## Licence

GPLv3+

## Credit

> 16:57 &lt;meornithorynque> et tu as besoin d'un nom ?<br>
> 16:57 &lt;meornithorynque> genre n'importe quoi ?<br>
> 16:57 &lt;meornithorynque> je propose "Prosopopée"

## Authors

By chronological order:

* [Bram](http://worlddomination.be), launched the project
* [Kload](https://github.com/Kloadut)
* [opi](https://github.com/opi)
* [taziden](https://www.libre-parcours.net/)
* [beudbeud](https://github.com/abeudin)
* [CapsLock](https://blog.legeox.net/)
* [Julien Malik](https://github.com/julienmalik)


CHANGELOG
CHANGELOG
=========

0.2 (2016-02-23)
----------------

* a lot new contributors stepped in, see https://github.com/Psycojoker/prosopopee#authors
* possibility to specify an (optional) menu https://github.com/Psycojoker/prosopopee#menu made by beudbeud
* configure Graphics Magick options on a global or per images fashion https://github.com/Psycojoker/prosopopee#gm and https://github.com/Psycojoker/prosopopee#images-handling by capslock and jmk
* support for various themes and a new material theme in addition of the default one https://github.com/Psycojoker/prosopopee#gm by beudbeud
* tags support in gallery settings https://github.com/Psycojoker/prosopopee#gallery-settingsyaml by beudbeud
* introduce internal CACHE format version to avoid breaking build when changing it

0.1 (2016-02-09)
----------------

* First pypi release


prosopopee/cache.py
import os
import json

CACHE_VERSION = 2


def remove_superficial_options(options):
    cleaned_options = options.copy()
    del cleaned_options["name"]
    if "text" in cleaned_options:
        del cleaned_options["text"]
    return cleaned_options


class Cache(object):
    cache_file_path = os.path.join(os.getcwd(), ".prosopopee_cache")

    def __init__(self, json):
        # fix: I need to keep a reference to json because for whatever reason
        # modules are set to None during python shutdown thus totally breaking
        # the __del__ call to save the cache
        # This wonderfully stupid behavior has been fixed in 3.4 (which nobody uses)
        self.json = json
        if os.path.exists(os.path.join(os.getcwd(), ".prosopopee_cache")):
            self.cache = json.load(open(self.cache_file_path, "r"))
        else:
            self.cache = {"version": CACHE_VERSION}

        if "version" not in self.cache or self.cache["version"] != CACHE_VERSION:
            print "info: cache format as changed, prune cache"
            self.cache = {"version": CACHE_VERSION}

    def needs_to_be_generated(self, source, target, options):
        if not os.path.exists(target):
            return True

        if target not in self.cache:
            return True

        cached_picture = self.cache[target]

        if cached_picture["size"] != os.path.getsize(source) or cached_picture["options"] != remove_superficial_options(options):
            return True

        return False

    def cache_picture(self, source, target, options):
        self.cache[target] = {"size": os.path.getsize(source), "options": remove_superficial_options(options)}

    def __del__(self):
        self.json.dump(self.cache, open(self.cache_file_path, "w"))


CACHE = Cache(json=json)


prosopopee/prosopopee.py
#!/usr/bin/env python

import os
import sys
import yaml
import shutil
import collections

from jinja2 import Environment, FileSystemLoader

from .cache import CACHE
from .utils import error

SETTINGS = {
    "show_date": True,
    "gm": {
        "quality": 75,
        "auto-orient": True,
        "strip": True,
        "resize": None
    }
}

class ListTranslateViewIterator(object):
  def __init__(self, lang, l):
    self.lang = lang
    self.it = iter(l)

  def next(self):
    return _(self.lang, self.it.next())

  def __iter__(self):
    return self

class ListTranslateView(collections.Sequence):
  def __init__(self, lang, l):
    self.lang=lang
    self.proxified_list = l

  def __len__(self):
    return len(self.proxified_list)

  def __getitem__(self, i):
    return _(self.lang, self.proxified_list[i])

  def __iter__(self):
    return ListTranslateViewIterator(self.lang, self.proxified_list)

  def __str__(self):
    # TODO: fails with unicode input, but we don't use it...
    return "[%s]" % (", ".join(["%s" % x for x in self]))

class DictTranslateViewIterator(object):
  def __init__(self, lang, d):
    self.lang = lang
    self.it = iter(d)

  def next(self):
    return _(self.lang, self.it.next())

  def __iter__(self):
    return self

class DictTranslateView(collections.Mapping):
  def __init__(self, lang, d):
    self.lang=lang
    self.proxified_dict = d

  def __len__(self):
    return len(self.proxified_dict)

  def __getitem__(self, key):
    if isinstance(self.proxified_dict[key], dict) and self.proxified_dict[key].has_key(self.lang):
      return self.proxified_dict[key][self.lang]
    else:
      return _(self.lang, self.proxified_dict[key])

  def __iter__(self):
    return DictTranslateViewIterator(self.lang, self.proxified_dict)

  def __str__(self):
    # TODO: fails with unicode input, but we don't use it...
    return "{%s}" % (", ".join(["'%s': '%s'" % (k,v) for (k,v) in self.iteritems()]))

def _(lang, o):
    # TODO: should support iterators too
    # TODO: testing for isinstance(o, collections.Mapping) creates an infinite recursion loop
    t = type(o)
    if isinstance(o, dict):
      return DictTranslateView(lang, o)
    elif isinstance(o, list):
      return ListTranslateView(lang, o)
    else:
      return o

class Image(object):
    base_dir = ""
    target_dir = ""

    def __init__(self, options):
        # assuming string
        if not isinstance(options, dict):
            options = {"name": options}
        self.options = SETTINGS["gm"].copy()  # used for caching, if it's modified -> regenerate
        self.options.update(options)

    @property
    def name(self):
        return self.options["name"]

    def gm(self, source, target, options):
        if CACHE.needs_to_be_generated(source, target, options):
            gm_switches = {
              "source": source,
              "target": target,
              "auto-orient" : "-auto-orient" if options["auto-orient"] else "",
              "strip": "-strip" if options["strip"] else "",
              "quality": "-quality %s" % options["quality"] if "quality" in options else "-define jpeg:preserve-settings",
              "resize": "-resize %s" % options["resize"] if options.get("resize", None) is not None else ""

            }
            command = "gm convert {source} {auto-orient} {strip} {quality} {resize} {target}".format(**gm_switches)
            print command
            os.system(command)
            CACHE.cache_picture(source, target, options)
        else:
            print "skipped %s since it's already generated (based on source unchanged size and images options set in your gallery's settings.yaml)" % target

    def copy(self):
        source, target = os.path.join(self.base_dir, self.name), os.path.join(self.target_dir, self.name)

        # XXX doing this DOESN'T improve perf at all (or something like 0.1%)
        # if os.path.exists(target) and os.path.getsize(source) == os.path.getsize(target):
            # print "Skipped %s since the file hasn't been modified based on file size" % source
            # return ""

        options = self.options.copy()
        if not options["auto-orient"] and not options["strip"]:
            shutil.copyfile(source, target)
            print source, "->", target
        else:
          # Do not consider quality settings here, since we aim to copy the input image
          # better to preserve input encoding setting
          del options["quality"]
          self.gm(source, target, options)
        return ""

    def generate_thumbnail(self, gm_geometry):
        thumbnail_name = self.name.split(".")
        thumbnail_name[-2] += "-%s" % gm_geometry
        thumbnail_name = ".".join(thumbnail_name)
        source, target = os.path.join(self.base_dir, self.name), os.path.join(self.target_dir, thumbnail_name)
        options = self.options.copy()
        options.update({"resize": gm_geometry})
        self.gm(source, target, options)
        return thumbnail_name

    def __repr__(self):
        return self.name


def main():
    if os.system("which gm > /dev/null") != 0:
        sys.stderr.write("ERROR: I can't locate the 'gm' binary, I won't be able to resize images, please install the 'graphicsmagick' package.\n")
        sys.exit(1)

    error(os.path.exists(os.path.join(os.getcwd(), "settings.yaml")), "I can't find a settings.yaml in the current working directory")

    settings = yaml.safe_load(open("settings.yaml", "r"))

    error(isinstance(settings, dict), "Your settings.yaml should be a dict")
    error(settings.get("title"), "You should specify a title in your main settings.yaml")

    if settings.get("settings", {}).get("gm"):
        SETTINGS["gm"].update(settings["settings"]["gm"])

    langs = settings.get("multilingual", [None])
    main_lang = langs[0]

    front_page_galleries_cover = []

    dirs = filter(lambda x: x not in (".", "..") and os.path.isdir(x) and os.path.exists(os.path.join(os.getcwd(), x, "settings.yaml")), os.listdir(os.getcwd()))

    error(dirs, "I can't find at least one directory with a settings.yaml in the current working directory (NOT the settings.yaml in your current directory, but one INSIDE A DIRECTORY in your current working directory), you don't have any gallery?")

    if not os.path.exists("build"):
        os.makedirs("build")

    theme = settings.get("settings", {}).get("theme", "exposure")

    error(os.path.exists(os.path.join(os.path.split(os.path.realpath(__file__))[0], "themes", theme)), "'%s' is not an existing theme, available themes are '%s'" % (theme, "', '".join(os.listdir(os.path.join(os.path.split(os.path.realpath(__file__))[0], "themes")))))

    templates = Environment(loader=FileSystemLoader([os.path.realpath(os.path.join(os.getcwd(), "templates")), os.path.join(os.path.split(os.path.realpath(__file__))[0], "themes", theme, "templates")]))
    index_template = templates.get_template("index.html")
    gallery_index_template = templates.get_template("gallery-index.html")
    page_template = templates.get_template("page.html")

    # XXX recursively merge directories
    if os.path.exists(os.path.join(os.getcwd(), "build", "static")):
        shutil.rmtree(os.path.join(os.getcwd(), "build", "static"))

    if os.path.exists(os.path.join(os.getcwd(), "static")):
        shutil.copytree(os.path.join(os.getcwd(), "static"), os.path.join(os.getcwd(), "build", "static"))
    else:
        shutil.copytree(os.path.join(os.path.split(os.path.realpath(__file__))[0], "themes", theme, "static"), os.path.join(os.getcwd(), "build", "static"))

    for gallery in dirs:
        gallery_settings = yaml.safe_load(open(os.path.join(os.getcwd(), gallery, "settings.yaml"), "r"))

        error(isinstance(gallery_settings, dict), "Your %s should be a dict" % (os.path.join(gallery, "settings.yaml")))
        error(gallery_settings.get("title"), "You should specify a title in %s" % (os.path.join(gallery, "settings.yaml")))

        if gallery_settings.get("public", True):
            error(gallery_settings.get("title"), "Your gallery describe in %s need to have a title" % (os.path.join(gallery, "settings.yaml")))
            error(gallery_settings.get("cover"), "You should specify a path to a cover picture in %s" % (os.path.join(gallery, "settings.yaml")))
            cover_image_path = os.path.join(gallery, gallery_settings["cover"])
            error(os.path.exists(cover_image_path), "File for %s cover image doesn't exist at %s" % (gallery, cover_image_path))

            front_page_galleries_cover.append({
                "title": gallery_settings["title"],
                "link": gallery,
                "sub_title": gallery_settings.get("sub_title", ""),
                "date": gallery_settings.get("date", ""),
                "tags": gallery_settings.get("tags", ""),
                "cover": cover_image_path,
            })

        if not os.path.exists(os.path.join("build", gallery)):
            os.makedirs(os.path.join("build", gallery))

        # this should probably be a factory
        Image.base_dir = os.path.join(os.getcwd(), gallery)
        Image.target_dir = os.path.join(os.getcwd(), "build", gallery)

        template_to_render = page_template if gallery_settings.get("static") else gallery_index_template
        for lang in langs:
            filename = "index.%s.html" % lang if lang and lang != main_lang else "index.html"
            open(os.path.join("build", gallery, filename), "w").write(template_to_render.render(settings=_(lang,settings), gallery=_(lang,gallery_settings), Image=Image).encode("Utf-8"))

    front_page_galleries_cover = list(reversed(sorted(front_page_galleries_cover, key=lambda x: x["date"])))

    # this should probably be a factory
    Image.base_dir = os.getcwd()
    Image.target_dir = os.path.join(os.getcwd(), "build")

    for lang in langs:
        filename = "index.%s.html" % lang if lang and lang != main_lang else "index.html"
        open(os.path.join("build", filename), "w").write(index_template.render(settings=_(lang,settings), galleries=_(lang,front_page_galleries_cover), Image=Image).encode("Utf-8"))

if __name__ == '__main__':
    main()
